text/0000-nested-publish.md

```diff
@@ -0,0 +1,198 @@
+* The sub-package is effectively “private” (in a sense like the Rust language's visibility system): it cannot be named as a dependency by any other package on `crates.io`, only by its parent package and sibling sub-packages. The parent package may still re-export items from it, or even the entire crate, in the same ways as it could do with a dependency on a normally published package.
+
+## Example: trait and derive macro
+
+Suppose we want to declare a trait-and-derive-macro package. We can do this as follows. The parent package would have this manifest `foo/Cargo.toml`:
+
+```toml
+[package]
+name = "foo"
+version = "0.1.0"
+edition = "2021"
+publish = true
+
+[dependencies]
+foo-macros = { path = "macros" }    # newly permitted
+```
+
+The sub-package manifest `foo/macros/Cargo.toml`:
+
+```toml
+[package]
+name = "macros"                     # this name need not be claimed on crates.io
+version = "0.1.0"                   # this version is not used for dependency resolution
+edition = "2021"
+publish = "nested"                  # new syntax
+
+[lib]
+proc-macro = true
+```
+
+Then you can `cargo publish` from within the parent `foo` directory, and this will create a single `foo` package on `crates.io`, with no `macros` (or `foo-macros`) package visible except when inspecting the source code or in compilation progress messages.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The following changes must be made across Cargo and `crates.io`:
+
+* **Manifest schema**
+    * The Cargo manifest now allows `"nested"` as a value for the `package.publish` key.
+* **`cargo package` &amp; `cargo publish`**
+    * Should refuse to publish a package if that package (not its sub-packages) has `publish = "nested"`.
+    * Exclude/include rules should, upon finding a sub-package, check if it is `publish = "nested"` and not automatically exclude it. Instead, they should treat it like any other subdirectory; in particular, it should be affected by explicitly specified exclude/include rules.
+    * Nested `Cargo.toml`s should be normalized in the same way the root `Cargo.toml` is, if they declare `publish = "nested"`, and not if they do not.
+        * This avoids modifying the publication behavior for existing packages, even if they contain project templates or invoke `cargo` to compile sub-packages to probe the behavior of the compiler.
+        * If the nested `Cargo.toml` has a syntax error such that its `package.publish` value cannot be determined, then if it is depended upon, emit an error; if it is not, emit a warning and do not normalize it.
+* **`crates.io`**
+    * Should allow `path` dependencies that were previously prohibited, at least provided that the named package in fact exists in the `.crate` archive file. The path must not contain any upward traversal (`../`) or other hazardous or non-portable components.
```

Comment:
if I'm looking at https://crates.io/crates/serde/1.0.174/dependencies, what should it show after this RFC?

I see multiple options:
- it just shows a "nested" dependency without any details
- it shows a "nested" dependency, whose details can then also be shown in some way
- it shows a union of the top-level package and nested dependencies

The first option is certainly the one with the least amount of work for the crates.io team, but it would have security implications since it would now be very easy to hide malicious dependencies in the subpackages.

The second option would require us to significantly refactor our data structures to support unversioned subpackages in some way. Given our backlog I'm not sure how viable that would be.

The third option could work, but then we'd have a significant difference between the `Cargo.toml` content and what the website shows, so it isn't a great solution either.

Are there other options how we could implement this on the registry side? Also, keep in mind that this would most likely have to be supported by the 3rd party registries too.

Reply:
> The third option could work, but then we'd have a significant difference between the Cargo.toml content and what the website shows, so it isn't a great solution either.

For me, from the users perspective the use of nested packages should be abstracted, so I would go with this option

> Are there other options how we could implement this on the registry side? Also, keep in mind that this would most likely have to be supported by the 3rd party registries too.

To clarify, I assume you mean if they provide a website with a similar dependency view as crates.io? 

I ask because we do need to keep in mind requirements for implementing a registry but an important distinction is minimum requirements vs additional clarifying features.

Reply:
> To clarify, I assume you mean if they provide a website with a similar dependency view as crates.io?

not necessarily. if we change the index layout in some way (see my other comment) then 3rd party registries would have to support that too.

Reply:
Yes, for index changes.  I was specifically talking in terms of the point raised here which seemed focused on the rendering.

Reply:
I've added discussion of the crates.io UI and index, and the lockfile, to the RFC text. I agree that for these purposes, nested package dependencies should probably be unioned into the parent package.